sound/soc/intel/boards/sof_ssp_amp.c

```diff
@@ -239,25 +168,24 @@ static int sof_ssp_amp_probe(struct platform_device *pdev)
 
 		if (mach->mach_params.codec_mask & IDISP_CODEC_MASK)
 			ctx->hdmi.idisp_codec = true;
-
-		sof_ssp_amp_card.num_links += ctx->hdmi_num;
 	} else {
 		ctx->hdmi_num = 0;
 	}
 
-	if (sof_ssp_amp_quirk & SOF_SSP_BT_OFFLOAD_PRESENT) {
+	if (sof_ssp_amp_quirk & SOF_SSP_BT_OFFLOAD_PRESENT)
 		ctx->bt_offload_present = true;
-		sof_ssp_amp_card.num_links++;
-	}
 
-	dai_links = sof_card_dai_links_create(&pdev->dev, ctx->amp_type,
-					      ctx->ssp_amp, ctx->dmic_be_num,
-					      ctx->hdmi_num,
-					      ctx->hdmi.idisp_codec);
-	if (!dai_links)
-		return -ENOMEM;
+	ctx->link_order_overwrite = SSP_AMP_LINK_ORDER;
 
-	sof_ssp_amp_card.dai_link = dai_links;
+	if (ctx->ssp_mask_hdmi_in) {
+		/* the topology supports HDMI-IN uses fixed BE ID for DAI links */
+		ctx->link_id_overwrite = SSP_AMP_LINK_IDS;
```

Comment:
the naming IS really really odd, I don't get the difference between  SSP_AMP_LINK_ORDER and SSP_AMP_LINK_IDS. The actual data is probably fine, it's just unreadable for the average reviewer :-)

Reply:
Would you suggest a suitable name please? My vocabulary is really limited...

Reply:
the main issue is to link the same with something that makes the reviewer think of HDMI RX with the SSP.

Reply:
SSP_AMP_DEFAULT_LINK_ORDER and SSP_AMP_WITH_HDMI_RX_LINK_ORDER ?

Reply:
I can do that but SSP_AMP_LINK_ORDER stores the order of DAI links in link array and SSP_AMP_LINK_IDS stores the id for each link. Although we use same macro SOF_LINK_ORDER to build these two variables but they are irrelevant.

Reply:
sorry but then I've again completely missed the idea. It seems that you are using the link_id_overwrite for different purposes?

Reply:
Yes, ctx->link_id_overwrite is for the id of BE links, lile here the id for amplifier link is always 2 and id for dmic01 is always 3...

#define HDMI_IN_BE_ID		0
#define SPK_BE_ID		2
#define DMIC01_BE_ID		3
#define INTEL_HDMI_BE_ID	5
/* extra BE links to support no-hdmi-in boards */
#define DMIC16K_BE_ID		4
#define BT_OFFLOAD_BE_ID	8

#define SSP_AMP_LINK_IDS	SOF_LINK_ORDER(HDMI_IN_BE_ID, \
					SPK_BE_ID,            \
					DMIC01_BE_ID,         \
					DMIC16K_BE_ID,        \
					INTEL_HDMI_BE_ID,     \
					BT_OFFLOAD_BE_ID,     \
					0)

ctx->link_order_overwrite is the order in dai link array, here the first link is HDMI-IN and second link is amplifier...

#define SSP_AMP_LINK_ORDER	SOF_LINK_ORDER(SOF_LINK_HDMI_IN, \
					SOF_LINK_AMP,            \
					SOF_LINK_DMIC01,         \
					SOF_LINK_DMIC16K,        \
					SOF_LINK_IDISP_HDMI,     \
					SOF_LINK_BT_OFFLOAD,     \
					SOF_LINK_NONE)

It seems to me that the misleading comes from using same SOF_LINK_ORDER macro to compost both value? Maybe I should just create a set of macros for id overwrite purpose?

Reply:
Yes they could be used concurrently so machine driver could specify dai link order or id of each link or both.

The default dai link order is declared in sof_board_helpers.c and this order will be used if not overwritten by ctx->link_order_overwrite

```
/* DEFAULT_LINK_ORDER: the order used in sof_rt5682 */
#define DEFAULT_LINK_ORDER	SOF_LINK_ORDER(SOF_LINK_CODEC, \
					SOF_LINK_DMIC01,       \
					SOF_LINK_DMIC16K,      \
					SOF_LINK_IDISP_HDMI,   \
					SOF_LINK_AMP,          \
					SOF_LINK_BT_OFFLOAD,   \
					SOF_LINK_HDMI_IN)
```

The default behavior of assigning id to each BE link is using dai link array index starting from zero. However, machine driver could use ctx->link_order_overwrite to specify the id for each link.

Reply:
@plbossart Is my update making sense to you? Thanks.